File: Dynamic Programming/2098.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(int(1e6))

INF = int(1e9)
n = int(input())
adj = [list(map(int, input().split())) for _ in range(n)]
dp = [[-1]*(1 << n)for _ in range(n)]

def DFS(st, state):
  if state == (1 << n) - 1:
    if adj[st][0]:
      return adj[st][0]
    else:
      return INF
  
  if dp[st][state] != -1:
    return dp[st][state]

  dp[st][state] = INF
    
  for i in range(1, n):
   if adj[st][i] == 0 or state & (1 << i):
    continue

   dp[st][state] = min(dp[st][state], DFS(i, state | (1 << i)) + adj[st][i])

  return dp[st][state]

print(DFS(0,1))

File: Dynamic Programming/2098.cpp
#include <bits/stdc++.h>
using namespace std;

const int INF = 0x7f7f7f7f;
int adj[16][16];
int dp[16][1 << 16];
int n;

int dfs(int cur, int curBit) {
  if (curBit == (1 << n) - 1) {
    if (adj[cur][0])
      return adj[cur][0];
    else
      return INF;
  }
  if (dp[cur][curBit] != -1) return dp[cur][curBit];

  dp[cur][curBit] = INF;
  
  for (int i = 1; i < n; i++) {
    if (adj[cur][i] == 0 || curBit & (1 << i)) continue;
    dp[cur][curBit] =
        min(dp[cur][curBit], dfs(i, curBit | (1 << i)) + adj[cur][i]);
  }
  return dp[cur][curBit];
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  cin >> n;
  for (int i = 0; i < n; i++) {
    fill(dp[i], dp[i] + (1 << n), -1);
    for (int j = 0; j < n; j++) cin >> adj[i][j];
  }
  cout << dfs(0, 1);
}